src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs

```diff
@@ -1711,16 +1613,148 @@ private string ReadLineFromConsole(bool endOnTab, string initialContent, bool ca
+            {
+                s += ConsoleControl.ReadConsole(handle, initialContent.Length, inputBuffer, MaxInputLineLength, endOnTab, out keyState);
+                Dbg.Assert(s != null, "s should never be null");
+
+                if (s.Length == 0)
+                {
+                    result = ReadLineResult.endedOnBreak;
+                    s = null;
+
+                    if (calledFromPipeline)
+                    {
+                        // make sure that the pipeline that called us is stopped
+
+                        throw new PipelineStoppedException();
+                    }
+
+                    break;
+                }
+
+                if (s.EndsWith(Crlf, StringComparison.Ordinal))
+                {
+                    result = ReadLineResult.endedOnEnter;
+                    s = s.Remove(s.Length - Crlf.Length);
+                    break;
+                }
+
+                int i = s.IndexOf(Tab, StringComparison.Ordinal);
+
+                if (endOnTab && i != -1)
+                {
+                    // then the tab we found is the completion character.  bit 0x10 is set if the shift key was down
+                    // when the key was hit.
+
+                    if ((keyState & 0x10) == 0)
+                    {
+                        result = ReadLineResult.endedOnTab;
+                    }
+                    else if ((keyState & 0x10) > 0)
+                    {
+                        result = ReadLineResult.endedOnShiftTab;
+                    }
+                    else
+                    {
+                        // do nothing: leave the result state as it was. This is the circumstance when we've have to
+                        // do more than one iteration and the input ended on a tab or shift-tab, or the user hit
+                        // enter, or the user hit ctrl-c
+                    }
+
+                    // also clean up the screen -- if the cursor was positioned somewhere before the last character
+                    // in the input buffer, then the characters from the tab to the end of the buffer need to be
+                    // erased.
+                    int leftover = RawUI.LengthInBufferCells(s.Substring(i + 1));
+
+                    if (leftover > 0)
+                    {
+                        Coordinates c = RawUI.CursorPosition;
+
+                        // before cleaning up the screen, read the active screen buffer to retrieve the character that
+                        // is overridden by the tab
+                        char charUnderCursor = GetCharacterUnderCursor(c);
+
+                        Write(StringUtil.Padding(leftover));
+                        RawUI.CursorPosition = c;
+
+                        restOfLine = s[i] + (charUnderCursor + s.Substring(i + 1));
+                    }
+                    else
+                    {
+                        restOfLine += s[i];
+                    }
+
+                    s = s.Remove(i);
```

Comment:
After reading this method, it seems like it would be a lot clearer if we renamed the `s` variable to something like `inputString` (probably in the other one as well)

Reply:
I'm used to it. :-)